source/adapters/level_zero/command_buffer.cpp

```diff
@@ -1346,12 +1515,174 @@ ur_result_t createUserEvent(ur_exp_command_buffer_handle_t CommandBuffer,
 
   return UR_RESULT_SUCCESS;
 }
-} // namespace
 
-UR_APIEXPORT ur_result_t UR_APICALL urCommandBufferEnqueueExp(
-    ur_exp_command_buffer_handle_t CommandBuffer, ur_queue_handle_t UrQueue,
-    uint32_t NumEventsInWaitList, const ur_event_handle_t *EventWaitList,
-    ur_event_handle_t *Event) {
+ur_result_t createProfilingCmdList(
+    ur_queue_handle_legacy_t Queue,
+    ur_exp_command_buffer_handle_t CommandBuffer, uint32_t NumEventsInWaitList,
+    const ur_event_handle_t *EventWaitList, ur_event_handle_t &UserEvent,
+    ze_command_list_handle_t &ProfilingCommandList) {
+
+  ur_command_list_ptr_t ProfilingCommandListPtr{};
+  UR_CALL(Queue->Context->getAvailableCommandList(
+      Queue, ProfilingCommandListPtr, false, NumEventsInWaitList, EventWaitList,
+      false));
+  ProfilingCommandList = ProfilingCommandListPtr->first;
+
+  // Multiple submissions of a command buffer implies that we need to save
+  // the event timestamps before resubmiting the command buffer. We
+  // therefore copy the these timestamps in a dedicated USM memory section
+  // before completing the command buffer execution, and then attach this
+  // memory to the event returned to users to allow to allow the profiling
+  // engine to recover these timestamps.
+  command_buffer_profiling_t *Profiling = new command_buffer_profiling_t();
+
+  Profiling->NumEvents = CommandBuffer->ZeEventsList.size();
+  Profiling->Timestamps =
+      new ze_kernel_timestamp_result_t[Profiling->NumEvents];
+
+  ZE2UR_CALL(zeCommandListAppendQueryKernelTimestamps,
+             (ProfilingCommandList, CommandBuffer->ZeEventsList.size(),
+              CommandBuffer->ZeEventsList.data(), (void *)Profiling->Timestamps,
+              0, nullptr, 1, &(CommandBuffer->ExecutionFinishedEvent->ZeEvent)));
+
+  UserEvent->CommandData = static_cast<void *>(Profiling);
+
+  return UR_RESULT_SUCCESS;
+}
+
+/**
+ * TODO
+ * This list submission is user facing. Needs to signal an event and associate
+ * it with the immediate command list helper. This needs to be done even if the
+ * user did not submit and event to urCommandBufferEnqueue()
+ * @param Queue
+ * @param CommandLists
+ * @param containsFinalCommandList
+ * @param SignalEvent
+ * @param isHostVisibleEvent
+ * @param NumEventsInWaitList
+ * @param EventWaitList
+ * @param useCopyEngine
+ * @return
+ */
+ur_result_t submitComputeEngineCmdLists(
+    ur_queue_handle_legacy_t Queue,
+    ur_exp_command_buffer_handle_t CommandBuffer, uint32_t NumEventsInWaitList,
+    const ur_event_handle_t *EventWaitList, ur_event_handle_t *UserEvent) {
+
+  std::vector<ze_command_list_handle_t> ComputeCommandLists{
+      CommandBuffer->ZeCommandListResetEvents,
+      CommandBuffer->ZeComputeCommandList};
```

Comment:
Wonder if this should this be `ZeComputeCommandListTranslated` if we're using an extension function, although it's a parameter not the command-list being executed on.

Reply:
Still unclear to me if the handles need to be translated for this specific entrypoint. Will do some more investigation into this.

Reply:
Confirmed that this is not needed since the entrypoint doesn't rely on zeDriverGetExtensionFunctionAddress